src/datasource/niftizarr/frontend.ts

```diff
@@ -0,0 +1,205 @@
+import type {
+  CompleteUrlOptions,
+  DataSource,
+  GetDataSourceOptions,
+} from "#src/datasource/index.js";
+import { DataSourceProvider } from "#src/datasource/index.js";
+import "#src/datasource/zarr/codec/bytes/resolve.js";
+import "#src/datasource/zarr/codec/crc32c/resolve.js";
+import "#src/datasource/zarr/codec/gzip/resolve.js";
+import "#src/datasource/zarr/codec/sharding_indexed/resolve.js";
+import "#src/datasource/zarr/codec/transpose/resolve.js";
+import {
+  parseDimensionSeparator,
+} from "#src/datasource/zarr/metadata/parse.js";
+import { parseOmeMetadata } from "#src/datasource/zarr/ome.js";
+import { completeHttpPath } from "#src/util/http_path_completion.js";
+import {
+  parseQueryStringParameters,
+  verifyObject,
+  verifyOptionalObjectProperty,
+} from "#src/util/json.js";
+import {
+  parseSpecialUrl,
+} from "#src/util/special_protocol_request.js";
+import { uncancelableToken } from "#src/util/cancellation.js";
+
+
+export class NiftiZarrDataSource extends DataSourceProvider {
+  constructor(public zarrVersion: 2 | 3 | undefined = undefined) {
+    super();
+  }
+  get description() {
+    const versionStr =
+      this.zarrVersion === undefined ? "" : ` v${this.zarrVersion}`;
+    return `Nifti Zarr${versionStr} data source`;
+  }
+  get(options: GetDataSourceOptions): Promise<DataSource> {
+    // Pattern is infallible.
+    let [, providerUrl, query] =
+      options.providerUrl.match(/([^?]*)(?:\?(.*))?$/)!;
+    const parameters = parseQueryStringParameters(query || "");
+    verifyObject(parameters);
+    const dimensionSeparator = verifyOptionalObjectProperty(
+      parameters,
+      "dimension_separator",
+      parseDimensionSeparator,
+    );
+    if (providerUrl.endsWith("/")) {
+      providerUrl = providerUrl.substring(0, providerUrl.length - 1);
+    }
+    return options.chunkManager.memoize.getUncounted(
+      { type: "nifti/getVolume", providerUrl },
+      /*{
+        type: "zarr:MultiscaleVolumeChunkSource",
+        providerUrl,
+        dimensionSeparator,
+      },*/
+      async () => {
+        const { url, credentialsProvider } = parseSpecialUrl(
+          providerUrl,
+          options.credentialsManager,
+        );
+        const metadata = await getMetadata(
+          options.chunkManager,
+          credentialsProvider,
+          url,
+          {
+            zarrVersion: this.zarrVersion,
+            explicitDimensionSeparator: dimensionSeparator,
+          },
+        );
+        if (metadata === undefined) {
+          throw new Error("No zarr metadata found");
+        }
+        let multiscaleInfo: ZarrMultiscaleInfo;
+        if (metadata.nodeType === "group") {
+          // May be an OME-zarr multiscale dataset.
+          const multiscale = parseOmeMetadata(url, metadata.userAttributes);
+          if (multiscale === undefined) {
+            throw new Error("Neithre array nor OME multiscale metadata found");
```

Comment:
Typo: 'Neither'

Reply:
good catch! this is something i copied - you can probably get yourself merged into google version if you wanted haha https://github.com/google/neuroglancer/blob/master/src/datasource/zarr/frontend.ts#L506